{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filtering primers against SARS-CoV-2\n",
    "A truncated list of primers from the validated [240,000 orthogonal primer library](https://www.pnas.org/doi/10.1073/pnas.0812506106) were checked against SARS-CoV-2 genome using BLAST. The generated alignment XML file were then parsed and further filtered to have a melting temperature between 60 and 65 degrees Celcius to generate a primer list (`seq_list.txt`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "\n",
    "fasta_file = 'bc25mer.240k.700trunct.fasta'\n",
    "xml_file = '7P4JPJ2M016-Alignment.xml'\n",
    "\n",
    "q_dict = SeqIO.index(fasta_file,'fasta')\n",
    "\n",
    "hits = []\n",
    "for record in NCBIXML.parse(open(xml_file)):\n",
    "    if record.alignments:\n",
    "        hits.append(record.query.split()[0])\n",
    "\n",
    "misses = set(q_dict.keys()) - set(hits)\n",
    "orphan_records = [q_dict[name] for name in misses]\n",
    "\n",
    "seq_file = open('seq_file.txt','w')\n",
    "for name in misses:\n",
    "    tm = mt.Tm_NN(q_dict[name].seq, Mg = 1.5)\n",
    "    if tm > 60 and tm < 65:\n",
    "        seq_file.write(str(q_dict[name].seq) + '\\n')\n",
    "seq_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Concatenate primers to create unique barcodes\n",
    "Primers are selected from the filtered primer list to create unique barcodes with the following structure:\n",
    "\n",
    "`Master forward + NNNNNN + Unique barcode + NNNNN + Master reverse`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "import string\n",
    "import xlsxwriter\n",
    "\n",
    "with open('seq_file.txt') as f:\n",
    "    seq_list = [line.strip() for line in f]\n",
    "           \n",
    "# Select master primers\n",
    "for_primer = seq_list[0]\n",
    "seq_list.remove(seq_list[0])\n",
    "rev_primer = seq_list[0]\n",
    "seq_list.remove(seq_list[0])\n",
    "\n",
    "# Generate internal barcodes\n",
    "barcodes_label = [('>barcode%i' % i) for i in range(96)]\n",
    "barcodes_alignment = [barcode for barcode in seq_list[:len(barcodes_label)]]\n",
    "barcodes_seq = [(for_primer + 'NNNNN' + barcode + 'NNNNN' + rev_primer) for barcode in seq_list[:len(barcodes_label)]]\n",
    "\n",
    "# Write FASTA file\n",
    "with open('templates_full.fasta', 'w') as f:\n",
    "    [f.write(line1 + '\\n' + line2 + '\\n' + '\\n') for (line1, line2) in zip(barcodes_label, barcodes_seq)]\n",
    "    \n",
    "with open('templates.fasta', 'w') as f:\n",
    "    [f.write(line1 + '\\n' + line2 + '\\n' + '\\n') for (line1, line2) in zip(barcodes_label, barcodes_alignment)]\n",
    "    \n",
    "# Generate xlsx file for IDT ordering\n",
    "row_label = [list(string.ascii_uppercase)[i] for i in range(8)]\n",
    "well_address = [i + str(j) for i in row_label for j in range(1, 13)]\n",
    "\n",
    "workbook = xlsxwriter.Workbook('IDT_order.xlsx')\n",
    "worksheet = workbook.add_worksheet()\n",
    "\n",
    "row = 0\n",
    "col = 0\n",
    "\n",
    "# Row headers\n",
    "row_headers = ['Well Address', 'Name', 'Sequences']\n",
    "[worksheet.write(row, col + i, row_headers[i]) for i in range(3)]\n",
    "\n",
    "idt_list = zip(well_address, barcodes_label, barcodes_seq)\n",
    "\n",
    "for row, idt_list in enumerate(idt_list):\n",
    "    worksheet.write_row(row + 1, 0, idt_list)\n",
    "workbook.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.9 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "6072b4658012f5d7b88a5a4d8b58603dae0bbd1763fc08f835dd5a424fca1cab"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
